To displayio.Display: expose fill_area and add a property for rotation

dastels opened the pull request.

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,17 +373,140 @@ const mp_obj_property_t displayio_display_bus_obj = {
+
+    displayio_display_fill_area(self, &area, mask, buffer);
+
+    mp_obj_array_t *result = array_new(BYTEARRAY_TYPECODE, buffer_size * 4);
```
Instead of allocating an array, please accept a bytearray or array in as an argument so that the buffer can be reused. Reusing the buffer will make it easier to allocate and reduce fragmentation.

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,17 +373,140 @@ const mp_obj_property_t displayio_display_bus_obj = {
+/*     .proxy = {(mp_obj_t)&displayio_display_get_screenshot_obj, */
+/*               (mp_obj_t)&mp_const_none_obj, */
+/*               (mp_obj_t)&mp_const_none_obj}, */
+/* }; */
```
Please remove this attribute.

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -337,6 +337,23 @@ const mp_obj_property_t displayio_display_height_obj = {
 
+//|   .. attribute:: rotation
+//|
+//|     The rotation of the display as an int in degrees.
```
Thanks for adding this!

dastels commented:
OK, Scott. I addressed the issues above. You now pass in an adequately large bytearray. Exceptions raised if it's not a bytearray or if it's not big enough.

deshipu commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,17 +373,132 @@ const mp_obj_property_t displayio_display_bus_obj = {
+    mp_int_t y = args[ARG_y].u_int;
+    mp_int_t w = args[ARG_width].u_int;
+    mp_int_t h = args[ARG_height].u_int;
+    mp_obj_array_t *result = (mp_obj_array_t *)(args[ARG_buffer].u_obj);
```
I would recommend using ``mp_get_buffer_raise`` to get the buffer and its size, then you don't need all those functions for accessing the array elements, etc.

deshipu replied:
You can see an example at https://github.com/adafruit/circuitpython/pull/2032/files#diff-28511b511b1f39aef03beef1a21a603eR82-R87

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,16 +373,95 @@ const mp_obj_property_t displayio_display_bus_obj = {
+
+//|   .. method:: fill_row(y, buffer)
+//|
+//|     Extract the pixels fro a single row
```
```suggestion
//|     Extract the pixels from a single row
```

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,16 +373,95 @@ const mp_obj_property_t displayio_display_bus_obj = {
+  enum { ARG_y, ARG_buffer };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_y, MP_ARG_INT | MP_ARG_KW_ONLY, {.u_int = -1} },
+        { MP_QSTR_buffer, MP_ARG_OBJ | MP_ARG_KW_ONLY, {} },
```
I'd remove both MP_ARG_KW_ONLYs and make them required positional arguments.

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,16 +373,95 @@ const mp_obj_property_t displayio_display_bus_obj = {
+
+    displayio_display_fill_area(self, &area, mask, buffer);
+
+    if ((result->len + result->free) >= (buffer_size * 4)) {
```
Please do this check before fill_area and then pass the given buffer straight into fill_area. That will simplify this implementation further.

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,16 +373,84 @@ const mp_obj_property_t displayio_display_bus_obj = {
+    }
+
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(result, &bufinfo, MP_BUFFER_WRITE);
```
Do this above where you have the array cast before the typecode check otherwise you are testing against memory that may not be what you think it is.

dastels commented:
@tannewt

tannewt commented on shared-bindings/displayio/Display.c:
```diff
@@ -356,16 +373,85 @@ const mp_obj_property_t displayio_display_bus_obj = {
+    mp_buffer_info_t bufinfo;
+    mp_get_buffer_raise(result, &bufinfo, MP_BUFFER_WRITE);
+
+    if (result->typecode != BYTEARRAY_TYPECODE) {
```
```suggestion
    if (bufinfo.typecode != BYTEARRAY_TYPECODE) {
```

dastels commented:
OK, @tannewt , travis is green.
